### PR TITLE
feat(ctb): introduce Types.sol

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Introduces a new library, Types.sol, meant to hold various different
structs that get used all over the place.

### CODE REVIEW

File: packages/contracts-bedrock/contracts/libraries/Types.sol
Comment (tynes) [inline packages/contracts-bedrock/contracts/libraries/Types.sol:47]: The nonce here is the nonce in the `L2ToL1MessagePasser` correct? Perhaps good to note so that it can be differentiated from a legacy tx. If solidity had type aliases then we could require that only a specific type of nonce is used here

File: packages/contracts-bedrock/contracts/test/Encoding.t.sol
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/Encoding.t.sol:18]: We need to delete this test because the encoding changed and now we have coverage via differential testing

File: packages/contracts-bedrock/contracts/test/Hashing.t.sol
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/Hashing.t.sol:26]: We need to delete this test, we have coverage via differential testing + the ci devnet deposit test covers this

**Comment by tynes** [comment]
Generally looks good to me, can approve after merge conflicts addressed

File: packages/contracts-bedrock/contracts/test/Hashing.t.sol
Comment (tynes) [reply packages/contracts-bedrock/contracts/test/Hashing.t.sol:26]: Note that the hashing scheme changed so this test should fail if you rebase

File: packages/contracts-bedrock/contracts/libraries/Types.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/libraries/Types.sol:47]: I think this would achieve what you're [looking for](https://docs.soliditylang.org/en/latest/types.html?highlight=custom%20types#user-defined-value-types)?

File: packages/contracts-bedrock/contracts/test/Encoding.t.sol
Comment (tynes) [reply packages/contracts-bedrock/contracts/test/Encoding.t.sol:18]: This test will break once this is rebased, i'm ok with re-adding it in the future

**Review by tynes — APPROVED** [review]

File: packages/contracts-bedrock/contracts/libraries/Types.sol
Comment (tynes) [reply packages/contracts-bedrock/contracts/libraries/Types.sol:47]: > I think this would achieve what you're [looking for](https://docs.soliditylang.org/en/latest/types.html?highlight=custom%20types#user-defined-value-types)?

Yes it could - I do think introducing type safety for certain interfaces is a good idea

**Review by maurelian — APPROVED** [review]

**Pull request merged**
